### ISSUE TITLE
REST API: product not saved when product is in more than one categories and position of product is updated via request - throws error "unable to save product"

### ISSUE THREAD

**Issue by roman204** [magento/magento2#19965]
### Preconditions (*)

1. take a vanilla magento (2.2.5)
2. try to save a product via REST API with category_links
the request url to save product is "/index.php/rest/all/V1/products/:sku"
```
...
"category_links": [
				{
					"position": 10,
					"category_id": "141"
				},
				{
					"position": 0,
					"category_id": "4"
				}

			],
...
```

### Steps to reproduce (*)

1. setup a vanilla magento2 (2.2.5) in my case
2. create the integration credentials for oauth
3. create a category unter the root category in magento (ID:3)
4. create a category unter the root category in magento (ID:4)
5. PUT product data via REST API ( with data:
```
{
	"product": {
		"sku": "my-sku",
		"name": "Lorem-ipsum",
		"attribute_set_id": 4,
		"price": 4.5,
		"status": 1,
		"visibility": 4,
		"type_id": "simple",
		"extension_attributes": {
			"website_ids": [
				0,
				1,
				2
			],
			"category_links": [
				{
					"position": 10,
					"category_id": "3"
				},
				{
					"position": 0,
					"category_id": "4"
				}
			]
		}
	}
}
```
6. change position of product in category 3 with request
```
{
	"product": {
		"sku": "my-sku",
		"name": "Lorem-ipsum",
		"attribute_set_id": 4,
		"price": 4.5,
		"status": 1,
		"visibility": 4,
		"type_id": "simple",
		"extension_attributes": {
			"website_ids": [
				0,
				1,
				2
			],
			"category_links": [
				{
					"position": 40,
					"category_id": "3"
				},
				{
					"position": 0,
					"category_id": "4"
				}
			]
		}
	}
}
```

### Expected result (*)

1. Position of product with sku: my-sku should be 40 in category 3

### Actual result (*)

1. ends with exeption: Undefined offset: 0 in /var/www/htdocs/vendor/magento/module-catalog/Model/ResourceModel/Product/CategoryLink.php on line 125" 
2. cause in file vendor/magento/module-catalog/Model/ResourceModel/Product/CategoryLink.php:125 
![image](https://user-images.githubusercontent.com/584168/50418863-24c63c80-082e-11e9-99b4-775618e8c258.png)

the "isset" check is missing as in file: vendor/magento/module-catalog/Model/Category/Link/SaveHandler.php
![image](https://user-images.githubusercontent.com/584168/50418845-0b24f500-082e-11e9-98ea-e45f1b9c315f.png)

screenshof of exception:
![image](https://user-images.githubusercontent.com/584168/50419052-0b71c000-082f-11e9-8fa1-1974db764759.png)

**Comment by magento-engcom-team** [comment]
Hi @roman204. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@roman204 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by roman204** [comment]
@magento-engcom-team give me 2.2.5 instance

**Comment by magento-engcom-team** [comment]
Hi @roman204. Thank you for your request. I'm working on Magento 2.2.5 instance for you

**Comment by magento-engcom-team** [comment]
Hi @roman204, here is your [Magento instance](https://i-19965-2-2-5.instances.magento-community.engineering).
Admin access: [https://i-19965-2-2-5.instances.magento-community.engineering/admin](https://i-19965-2-2-5.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by roman204** [comment]
seems to be fixed in 2.2-develop, i'll check it out

**Comment by roman204** [comment]
Seems to be the same behavior on 2.2.7. https://github.com/magento/magento2/blob/2.2.7/app/code/Magento/Catalog/Model/ResourceModel/Product/CategoryLink.php

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by roman204** [comment]
is fixed with commit 3fe6c70882669acf77838da7bf86f678e690d6ed and 6f6694c3b1ca869fc6bd238c68dc8772d2261471

**Comment by roman204** [comment]
my fault, only fixed in 2.2-develop

**Comment by ghost** [comment]
Hi @roman204 thank you for you report, this already fixed on 2.3-develop branch
![selection_249](https://user-images.githubusercontent.com/40756439/50439346-615f6a00-08fa-11e9-9651-4f86a8f3d8c8.png)

**Comment by roman204** [comment]
ok thanks, cause i need it now and currently the customer is on 2.2 i made a patch which can be executed on installation with composer: 
```
Index: vendor/magento/module-catalog/Model/ResourceModel/Product/CategoryLink.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- vendor/magento/module-catalog/Model/ResourceModel/Product/CategoryLink.php	(date 1545815826000)
+++ vendor/magento/module-catalog/Model/ResourceModel/Product/CategoryLink.php	(date 1545815826000)
@@ -114,16 +114,16 @@
     private function processCategoryLinks($newCategoryPositions, &$oldCategoryPositions)
     {
         $result = ['changed' => [], 'updated' => []];
+
+        $oldCategoryPositions = array_values($oldCategoryPositions);
+        $oldCategoryList = array_column($oldCategoryPositions, 'category_id');
         foreach ($newCategoryPositions as $newCategoryPosition) {
-            $key = array_search(
-                $newCategoryPosition['category_id'],
-                array_column($oldCategoryPositions, 'category_id')
-            );
+            $key = array_search($newCategoryPosition['category_id'], $oldCategoryList);
 
             if ($key === false) {
                 $result['changed'][] = $newCategoryPosition;
             } elseif ($oldCategoryPositions[$key]['position'] != $newCategoryPosition['position']) {
-                $result['updated'][] = $newCategoryPositions[$key];
+                $result['updated'][] = $newCategoryPosition;
                 unset($oldCategoryPositions[$key]);
             }
         }

```
